src/radical/pilot/agent/scheduler/base.py

```diff
@@ -477,133 +476,316 @@ def _schedule_units(self, units):
-          #                 unit['description']['gpu_processes'],
-          #                 pprint.pformat(unit['slots']))
+            # reclaim resources from completed tasks
+            a, r = self._unschedule_completed()
+            if r: resources  = True
+            if a: active    += 'u'
 
-        # True signals success
-        return True
+            if not active:
+                time.sleep(0.1)  # FIXME: configurable
 
 
     # --------------------------------------------------------------------------
     #
-    def _handle_cuda(self, unit):
+    def _schedule_waitlist(self):
 
-        # Check if unit requires GPUs.  If so, set CUDA_VISIBLE_DEVICES to the
-        # list of assigned  GPU IDs.  We only handle uniform GPU setting for
-        # now, and will isse a warning on non-uniform ones.
-        #
-        # The default setting is ``
-        #
-        # FIXME: This code should probably live elsewhere, not in this
-        #        performance critical scheduler base class
-        #
-        # FIXME: The specification for `CUDA_VISIBLE_DEVICES` is actually LM
-        #        dependent.  Assume the scheduler assigns the second GPU.
-        #        Manually, one would set `CVD=1`.  That also holds for launch
-        #        methods like `fork` which leave GPU indexes unaltered.  Other
-        #        launch methods like `jsrun` mask the system GPUs and only the
-        #        second GPU is visible, at all, to the task.  To CUDA the system
-        #        now seems to have only one GPU, and we need to be set `CVD=0`.
-        #
-        #        In other words, CVD sometimes needs to be set to the physical
-        #        GPU IDs, and at other times to the logical GPU IDs (IDs as
-        #        visible to the task).  This also implies that this code should
-        #        actually live within the launch method.  On the upside, the LM
-        #        should also be able to handle heterogeneus tasks.
+        self.slot_status("before schedule waitlist")
+
+        # cycle through waitlist, and see if we get anything placed now.
         #
-        #        For now, we hardcode the CVD ID mode to `logical`, thus
-        #        assuming that unassigned GPUs are masked away, as for example
-        #        with `jsrun`.
-        cvd_id_mode = 'logical'
-
-        unit['description']['environment']['CUDA_VISIBLE_DEVICES'] = ''
-        gpu_maps = list()
-        for node in unit['slots']['nodes']:
-            if node['gpu_map'] not in gpu_maps:
-                gpu_maps.append(node['gpu_map'])
-
-        if not gpu_maps:
-            # no gpu maps, nothing to do
+        # sort by inverse tuple size to place larger tasks first and backfill
+        # with smaller tasks.  We only look at cores right now - this needs
+        # fixing for GPU dominated loads.
+
+        self._waitpool.sort(key=lambda x: x['tuple_size'][0], reverse=True)
+        scheduled, unscheduled = ru.lazy_bisect(self._waitpool,
```

Comment:
`_waitpool` starts with blank, does lazy_bisect() work okay with the empty list?

Reply:
Nice catch, I'll add a test in RU :-)